# Coding Style PEP-8
- imports at the top each in tis own line
- indent code using space instead of tabs
- use for spaces for each  indentation levels
- limit lines to 79 chars (72 for comments and docstrings)
- seperate functions and classe by two blank lines
- in classes one blank line seperates methods
- no white spce around functions calls, indexes and keyword arguments

Whitespace:
- *spam(fn)*  is better than *spam ( fn )*
- *list[index]* is better than *list [index]*

https://peps.python.org/pep-0008/#whitespace-in-expressions-and-statements

# documentation .\_\_doc\_\_


In [31]:
# doc of a function
print(map.__doc__)

map(func, *iterables) --> map object

Make an iterator that computes the function using arguments from
each of the iterables.  Stops when the shortest iterable is exhausted.


In [32]:
import collections
# doc of a module
print(collections.__doc__)

This module implements specialized container datatypes providing
alternatives to Python's general purpose built-in containers, dict,
list, set, and tuple.

* namedtuple   factory function for creating tuple subclasses with named fields
* deque        list-like container with fast appends and pops on either end
* ChainMap     dict-like class for creating a single view of multiple mappings
* Counter      dict subclass for counting hashable objects
* OrderedDict  dict subclass that remembers the order entries were added
* defaultdict  dict subclass that calls a factory function to supply missing values
* UserDict     wrapper around dictionary objects for easier dict subclassing
* UserList     wrapper around list objects for easier list subclassing
* UserString   wrapper around string objects for easier string subclassing




# docstring
- always in triple quotes
- first line should summarize the functionality
- Modules: list important classes, function, exceptions
- Classes: list important methods, enums, static member fct.
- Functions: 
  - list params and explain in it, one per line
  - if there is a return value, list it, otherwise omit
  - list exceptions raised by the fct
https://docutils.sourceforge.io/docs/peps/pep-0257.html

In [34]:
def myFunction(arg1, arg2=None):
    """myFunction(arg1, arg2=None) --> doesn't really do anything
    
    Paramaters:
    arg1: first argument.
    arg2: second argument
    """
    print(arg1, arg2)


print(myFunction.__doc__)

myFunction(arg1, arg2=None) --> doesn't really do anything
    
    Paramaters:
    arg1: first argument.
    arg2: second argument
    


# magic commands
- Magic commands are enhancements to normal python code which are provided by the IPython kernel.
- They are usually prefixed by a “%” character
- % prefix represents that the command operates over a single line of code whereas 
- %% prefix allows the command to operate over an entire cell.


In [12]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %conda  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%

In [13]:
%magic


IPython's 'magic' functions

The magic function system provides a series of functions which allow you to
control the behavior of IPython itself, plus a lot of system-type
features. There are two kinds of magics, line-oriented and cell-oriented.

Line magics are prefixed with the % character and work much like OS
command-line calls: they get as an argument the rest of the line, where
arguments are passed without parentheses or quotes.  For example, this will
time the given statement::

        %timeit range(1000)

Cell magics are prefixed with a double %%, and they are functions that get as
an argument not only the rest of the line, but also the lines below it in a
separate argument.  These magics are called with two arguments: the rest of the
call line and the body of the cell, consisting of the lines below the first.
For example::

        %%timeit x = numpy.random.randn((100, 100))
        numpy.linalg.svd(x)

will time the execution of the numpy svd routine, running the assignment 

In [14]:
%quickref


IPython -- An enhanced Interactive Python - Quick Reference Card

obj?, obj??      : Get help, or more help for object (also works as
                   ?obj, ??obj).
?foo.*abc*       : List names in 'foo' containing 'abc' in them.
%magic           : Information about IPython's 'magic' % functions.

Magic functions are prefixed by % or %%, and typically take their arguments
without parentheses, quotes or even commas for convenience.  Line magics take a
single % and cell magics are prefixed with two %%.

Example magic function calls:

%alias d ls -F   : 'd' is now an alias for 'ls -F'
alias d ls -F    : Works if 'alias' not a python name
alist = %alias   : Get list of aliases to 'alist'
cd /usr/share    : Obvious. cd -<tab> to choose from visited dirs.
%cd??            : See help AND source for magic %cd
%timeit x=10     : time the 'x=10' statement with high precision.
%%timeit x=2**100
x**100           : time 'x**100' with a setup of 'x=2**100'; setup code is not
                   co

In [17]:
# writes the whole block into a file 
%%writefile myfile.py 

def myprint():
    print("I'm a magic file")

myprint()

Overwriting myfile.py


In [19]:
# runs the external file from notebook cell
# %run also allows the execute of external jupyter notebooks
%run myfile.py 

I'm a magic file


In [27]:
# %pycat allows you to display the content of any file in any directory.
%pycat myfile.py


def myprint():
    print("I'm a magic file")

myprint()


In [28]:
%who #displays all the variables used in the whole notebook

a	 i	 myprint	 os	 sys	 


In [32]:
b = 'string'
%who str #In order to dsiplay specific data type variables, we need to pass the data type after the magic command.

b	 


In [25]:
%%time
a = []
for i in range(1000):
    a.append(i**2)


CPU times: user 394 µs, sys: 43 µs, total: 437 µs
Wall time: 439 µs


In [34]:
%% html # allows us to write html code in the cell

<html>
<body>
<table>
        <tr> 
            <th>Name</th> 
            <th>Country</th> 
            <th>Age</th> 
        </tr> 
        <tr> 
            <td>Sid</td> 
            <td>India</td> 
            <td>22</td> 
        </tr>
        <tr> 
            <td>Dave</td> 
            <td>UK</td> 
            <td>28</td> 
        </tr>
</table>
</body>
</html>

UsageError: Cell magic `%%` not found.


# Truth Values
Boolean True
- any object in Python is considered to be equivalent to boolean True unless it's call defines a bool that returns False or retruns a len(x)=0

Built-in objects that evaluate to False
- False = None = 0 = 0j = Decimal(0) = Fraction (0, x)
- Empty sequences and collections are also false (), [], {}
- Empty sets and ranges amount to false set(), range(0)

AND, OR, NOT
- are short curcuit operators they check just the necessary operands
- AND (False AND True) chescks the first operant, if it is False the second operant is not checked
- OR only checks the second operand if the first one is False

# Strings vs. Bytes
- strings and bytes are not directly interchangeable
- strings contain unicode
- bytes are raw 8-bit values
- decode transforms bytes to a string
- encode transforms strings in bytes
  - letters, punctuation, symbols, whitespace, and control characters become integers and ultimately bits. 
  - Each character can be encoded to a unique sequence of bits.

In [2]:
# The results of str.encode() is a bytes object.
# Both bytes literals (such as b"r\xc3\xa9sum\xc3\xa9") and the representations of bytes permit only ASCII characters.
encoded = "résumé".encode("utf-8")
print(encoded)

decoded = b"r\xc3\xa9sum\xc3\xa9".decode("utf-8")
print(decoded)

b'r\xc3\xa9sum\xc3\xa9'
résumé


In [3]:
# to combine bytes and strings one needs to be en- or decoded
b = bytes([0x41, 0x42, 0x43, 0x44])  # asci letters
s = 'this is a string'

s2 = b.decode('utf-8')
print(s + s2)

this is a stringABCD


In [12]:
# hexadecimal
b3 = s.encode('utf-32')
b3

b'\xff\xfe\x00\x00t\x00\x00\x00h\x00\x00\x00i\x00\x00\x00s\x00\x00\x00 \x00\x00\x00i\x00\x00\x00s\x00\x00\x00 \x00\x00\x00a\x00\x00\x00 \x00\x00\x00s\x00\x00\x00t\x00\x00\x00r\x00\x00\x00i\x00\x00\x00n\x00\x00\x00g\x00\x00\x00'

# template strings
- value subsitution
- more secure than the simple format method

In [18]:
# usual formatting with format
str1 = 'This is {0} by {1}'.format('Python Advanced', 'Marco')
str1

'This is Python Advanced by Marco'

In [17]:
from string import Template
templ = Template('This is ${notebook} by ${author}')
str2 = templ.substitute(notebook='Python Advanced', author='Marco')
str2

'This is Python Advanced by Marco'

In [19]:
dict = {
    'author': 'Marco',
    'notebook': 'Python Advanced'
}

str3 = templ.substitute(dict)
str3

'This is Python Advanced by Marco'

# Functions
- define a function - A function is a block of code, which only runs when it is called.
- Arguments are specified after the function name, inside the parentheses.

In [3]:
def greet_user(fname, lname):            # parameters are the placeholders
    print(f"Hi,  {fname}, {lname}")         # argument is the definite value given to the function as a parameter
    print("Welcome")
    
greet_user("Hans", "Z")             # positional arguments
greet_user(lname="Z", fname="Hans") # keyword arguments

Hi,  Hans, Z
Welcome
Hi,  Hans, Z
Welcome


## return 
- You can use the return statement to make your functions send Python objects back to the caller code.
- every function returns a NONE when there is no return statement
- return stops the functions execution

In [5]:
def square(number):
    return print(number*number)

square(3)

9


## typing
- In the following function, the argument name is expected to be of type str
- and the return type shold be int. Subtypes are accepted as arguments.

In [63]:
def name_len(name: str) -> int:
    return len(name)

name_len("Marco")

5

## arguments (args)
- **Parameters** are defined by the names that appear in a function definition, whereas **arguments** are the values actually passed to a function when calling it. 
- Parameters define what types of arguments a function can accept. 
- A __parameter__ is the variable listed inside the parentheses in the function definition.
- An __argument__ is the value that are sent to the function when it is called.

#### Parameters
Parameters are part of a function definition (def or lambda). There are five different kinds of parameters:

- **positional-or-keyword**: 
Normal parameters in a function definition, with or without default values. 
Each has a name and an index, and can accept a positional argument with the same index, 
or a keyword argument with the same name, or (if it has a default value) nothing. 
Technically, every parameter before the first bare *, var-positional, or var-keyword is a positional-or-keyword parameter.
- **positional-only**: 
Only found in builtin/extension functions. 
Each has a name and an index, but only accepts positional arguments with the same index.
- **var-positional**: 
This is the *args.
This accepts a sequence consisting of all positional arguments whose index is larger 
than any positional-or-keyword or positional-only parameter.
(Note that you can also specify a bare * here. In that case, you don't take variable positional arguments.
You do this to set off keyword-only from positional-or-keyword parameters.)
- **keyword-only**: 
These are parameters that come after a * or *args, with or without default values.
Each has a name only, and accepts only keyword arguments with the same name. 
Technically, every parameter after the first bare * or var-positional,
but before the var-keyword (if any), is a keyword-only parameter.
- **var-keyword**: 
This is the **args.
This accepts a mapping consisting of all keyword arguments whose name does not 
match any positional-or-keyword or keyword-only parameter.

#### Arguments
Arguments are part of a function call. There are four different kinds of arguments:

- **positional**: Arguments without a name.
        Each is matched to the positional-or-keyword or positional-only parameter with the same index, or to the var-positional parameter if there is no matching index (or, if there is no var-positional parameter, it's an error if there is no match).
- **keyword**: Arguments with a name.
        Each is matched to the postional-or-keyword or keyword-only parameter with the same name, or to the var-keyword parameter if there is no matching name (or, if there is no var-keyword parameter, it's an error if there is no match).
- **packed positional**: An iterator preceded by *.
        The iterator is unpacked, and the values treated as separate positional arguments.
- **packed keyword**: A mapping preceded by **.
        The mapping is iterated, and the key-value pairs treated as separate keyword arguments.

There is no direct connection between parameters with a default value and keyword arguments. You can pass keyword arguments to parameters without default values, or positional arguments to parameters with them.


### positional args
-  By default, a function must be called with the correct number of arguments. 
-  the args passed are assigned according to their position

In [113]:
def my_function(fname, lname):
  print(fname + " " + lname)

my_function("Emil", "Hauser") 

Emil Hauser


### keyword args
- Keyword Arguments (kwargs)
-  You can also send arguments with the 'key = value' syntax.
- This way the order of the arguments does not matter.

In [ ]:
def my_function(child3, child2, child1):
  print("The youngest child is " + child3)

my_function(child1 = "Emil", child2 = "Tobias", child3 = "Linus")

#### keyword-only Arguments
- if we want the function caller to explicitly use a keyword 
- the * here means ‘don’t allow any positional arguments beyond this point’
- otherwise any further (accidentially) added parameter is taken as the keyword parameter

In [34]:
def k_only(arg1, arg2, *,  sub=False):
    return print(arg1, arg2, sub)

k_only(1, 2, sub=3)

1 2 3


### *args
-  *args, Variable Arguments
-  *args gets all the excess positional arguments
-  If you do not know how many arguments will be passed into your function
-  This way the function will receive a tuple of arguments, and can access the items accordingly
- good when # of args is small
- add flexibility to the fct.


In [ ]:
def my_function(*kids):
  print("The youngest child is " + kids[2])

my_function("Emil", "Tobias", "Linus")

In [15]:
# if we add a positional param like this addition(base, *args)
# the fct. is off bc it takes base and misses the last value in the *args
def addition(*args):
    result=0
    for arg in args:
        result += arg
    return result

print(addition(2, 4, 5))
print(addition(10 ,20, 4, 81, 41))
nums = [1, 4, 6, 6, 10]
print(addition(*nums))

11
156
27


### kwargs
- var-keyword parameter/ keyword dictionary parameter slot
- **kwargs, variable Keyword Arguments
- **kwargs gets all the excess keyword arguments
- If the number of keyword arguments is unknown, add a double ** before the parameter name.


In [ ]:
def my_function(**kid):
  print("His last name is " + kid["lname"])

my_function(fname = "Tobias", lname = "Refsnes")

In [ ]:
# default argument
def my_function(country = "Norway"):
  print("I am from " + country)

my_function("India")
my_function()

## higher order function 
- contains other functions as a parameter or returns a function as an output

In [68]:
def shout(text):
    return text.upper()

print(shout('Hello'))


HELLO


## Passing a Function as an argument to another function

In [73]:
def shout(text):
    return text.upper()

def whisper(text):
    return text.lower()

def greet(func):
    # storing the function in a variable
    greeting = func("Hi, I am created by a function passed as an argument.")
    print(greeting)

greet(shout)  # function greet which takes a function as an argument
greet(whisper)

HI, I AM CREATED BY A FUNCTION PASSED AS AN ARGUMENT.
hi, i am created by a function passed as an argument.


## recursion

In [ ]:
# recursion - means a defined function can call itself
def tri_recursion(k):
  if(k > 0):
    result = k + tri_recursion(k - 1)
    print(f"k: {k}, k-1: {k-1}, result: {result}")
  else:
    result = 0
  return result
print("\nRecursion Example Results")
x = tri_recursion(2)
print(x)


Recursion Example Results
k: 1, k-1: 0, result: 1
k: 2, k-1: 1, result: 3
3


## Assigning function to a variable

In [70]:
yell = shout
print(yell('Hello'))

HELLO


## lambda function
- lambda function is a small anonymous fct.
- they don't have names
- can be passed as arguments to other fct.
- can take any number of arguments, but has only one expression
- lambda(parameters) : (expression)

In [ ]:
x = lambda a, b : a ** b           # two args followed by clause
print(x(2, 3))                              # x is return

8


In [23]:
ctemp =[0, 12, 34, 100]
#  temp conversion is a very mall fct
# we can easliy put it in a lambda
list(map(lambda t: (t*9/5)+32, ctemp))

[32.0, 53.6, 93.2, 212.0]

In [ ]:
# power of lambda is better shown when you use them 
# as an anonymous function inside another function
def times(n):
  return lambda a : a ** n

squared = times(2)
print(squared(11))

power_4 = times(4)
print(power_4(2))

121
16


In [ ]:
# Example of lambda function using if-else
max = lambda a, b : a if(a > b) else b
print(max(1, 2))

2


In [ ]:
# lambda - inner and outer function
List = [[2,3,4],[1, 4, 16, 64],[3, 6, 9, 12]]
sortList = lambda x: (sorted(i) for i in x)                     # Sort each sublist - inner fct.
secondLargest = lambda l, sL : [i[-2] for i in sL(l)]     # Get the second largest element - outer fct.
res = secondLargest(List, sortList)
print(res)

[3, 16, 9]


# map, reduce, filter

## map()
- map() returns a map object (which is an iterator) of the results after
- applying the given function to each item of a given iterable (list, tuple etc.)
- The map() function executes a specified function for each item in an iterable.
-  The item is sent to the function as a parameter.

In [47]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
final_list = list(map(lambda x: x**2, li))
print(final_list)


[25, 49, 484, 9409, 2916, 3844, 5929, 529, 5329, 3721]


In [1]:
def toGrades(x):
    if x > 90: return 'A'
    elif 80 <= x & x < 90: return 'B'
    elif 70 <= x & x <80: return 'C'
    elif 65 <= x & x <70: return 'D'
    else: return 'F'

In [6]:
grades = (102, 99, 86, 67, 50, 81, 76, 60)
grade_sort = sorted(grades)
# maps letters to numbers according to the function
letters = list(map(toGrades, grade_sort))
print(grade_sort)
print(letters)

[50, 60, 67, 76, 81, 86, 99, 102]
['F', 'F', 'D', 'C', 'B', 'B', 'A', 'A']


In [ ]:
numbers = (1, 2, 3, 4)
scale = lambda a: a*100
result = map(scale, numbers) # takes each item of numbers applies the fct. and returns a list
list(result)

[100, 200, 300, 400]

In [ ]:
len_func = lambda x: len(x)

x = list(map(len_func, ('apple', 'banana', 'cherry')))    # map creates an map object, thus convert it to a list
print(x)


[5, 6, 6]


## filter()
- filter() takes in a function and a list as arguments.
- __filters out all the elements of a sequence, for which the function returns True.__
- filter(function, iterable)

In [ ]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
final_list = list(filter(lambda x: (x%2 == 0) , li))
final_list

[22, 54, 62]

In [ ]:
ages = [13, 90, 17, 59, 21, 60, 5]
adults = list(filter(lambda age: age>18, ages))
print(adults)

[90, 59, 21, 60]


## reduce()
- needs to be imported as it resides in the functools module.
- performs a repetitive operation over the pairs of the iterable.
-  implements a technique called folding - reduce a list of items to a single cumulative value
- takes an existing function, applys it cumulatively to all the items in an iterable

In [ ]:
from functools import reduce
lis = [10, 32, 51, 76, 2]
maximum = reduce(lambda a,b : a if a > b else b,lis)
print (f"The maximum element of the list is : {maximum}")

The maximum element of the list is : 76


# built-in functions

In [ ]:
abs()	        # Returns the absolute value of a number
all()	          # Returns True if all items in an iterable object are true
any()	        # Returns True if any item in an iterable object is true
ascii()	        # Returns a readable version of an object. Replaces none-ascii characters with escape character
bin()	        # Returns the binary version of a number
bool()	            # Returns the boolean value of the specified object
bytearray()	    # Returns an array of bytes
bytes()	            # Returns a bytes object
callable()	              # Returns True if the specified object is callable, otherwise False
chr()	                     # Returns a character from the specified Unicode code.
ord()	                    # Convert an integer representing the Unicode of the specified character
classmethod()	    # Converts a method into a class method
compile()	            # Returns the specified source as an object, ready to be executed
complex()	           # Returns a complex number
delattr()	              # Deletes the specified attribute (property or method) from the specified object
dict()	                    # Returns a dictionary (Array)
dir()	                     # Returns a list of the specified object's properties and methods
divmod()	           # Returns the quotient and the remainder when argument1 is divided by argument2
enumerate()	        # Takes a collection (e.g. a tuple) and returns it as an enumerate object
eval()	                   # Evaluates and executes an expression
exec()	                  # Executes the specified code (or object)
filter()	          # Use a filter function to exclude items in an iterable object
float()	             # Returns a floating point number
format()	      # Formats a specified value     
frozenset()	    # Returns a frozenset object

getattr()	      # Returns the value of the specified attribute (property or method)
setattr()	       # Sets an attribute (property/method) of an object
hasattr()	      # Returns True if the specified object has the specified attribute (property/method)
isinstance()	             # Returns True if a specified object is an instance of a specified object
issubclass()	             # Returns True if a specified class is a subclass of a specified object

globals()	      # Returns the current global symbol table as a dictionary
hash()	            # Returns the hash value of a specified object
help()	            # Executes the built-in help system
hex()	            # Converts a number into a hexadecimal value
id()	              # Returns the id of an object
input()	           # Allowing user input
int()	             # Returns an integer number
iter()	             # Returns an iterator object
len()	             # Returns the length of an object
list()	              # Returns a list
locals()	        # Returns an updated dictionary of the current local symbol table
map()	            # Returns the specified iterator with the specified function applied to each item
max()	            # Returns the largest item in an iterable
memoryview()	             # Returns a memory view object
min()	             # Returns the smallest item in an iterable
next()	             # Returns the next item in an iterable
object()	       # Returns a new object
oct()	              # Converts a number into an octal
open()	           # Opens a file and returns a file object
pow()	            # Returns the value of x to the power of y
print()	             # Prints to the standard output device
property()	             # Gets, sets, deletes a property
range()	            # Returns a sequence of numbers, starting from 0 and increments by 1 (by default)
repr()	             # Returns a readable version of an object
reversed()	             # Returns a reversed iterator
round()	          # Rounds a numbers
set()	             # Returns a new set object
slice()	             # Returns a slice object
sorted()	      # Returns a sorted list
staticmethod()	             # Converts a method into a static method
str()	                # Returns a string object
sum()	             # Sums the items of an iterator
super()	            # Returns an object that represents the parent class
tuple()	             # Returns a tuple
type()	             # Returns the type of an object
vars()	             # Returns the __dict__ property of an object
zip()	              # Returns an iterator, from two or more iterators

In [21]:
list1=[1, 2, 3, 0, 5, 6]
# 'any' goes through the list and returns True if any of the values evaluates to True
print(any(list1))
print(all(list1)) # are all True?


True
False


# Comprehensions

## list comprehension

In [3]:
# a list of converted elements can be produced like this or...
def FahrenheitToCelsius(t):
    return (t * 9 / 5) + 32

list(map(FahrenheitToCelsius, [32, 65, 104, 212]))

[89.6, 149.0, 219.2, 413.6]

In [2]:
# ... much shorter with list comprehension
[(t * 9 / 5) + 32 for t in [32, 65, 104, 212]]

[89.6, 149.0, 219.2, 413.6]

In [7]:
evens = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
list(map(lambda e: e**2, evens)) 
[e**2 for e in evens] # the same but shorter

[4, 16, 36, 64, 100, 144, 196, 256, 324, 400]

## comprehension with predicate expression / condition

In [8]:
odds = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
[e**2 for e in odds if e>3 and e <17] 

[25, 49, 81, 121, 169, 225]

## dictionary comprehension

In [10]:
ctemps =[0, 12, 34, 100]
tempDict={t: (t*9/5) +32 for t in ctemps if t<100}
print(tempDict)
print(tempDict[12])

{0: 32.0, 12: 53.6, 34: 93.2}
53.6


In [13]:
team1 = {"Jones": 24, "Jameson": 18, "Smith": 58, "Burns": 7}
team2={"White": 12, "Macke": 88, "Pierce": 4}

# do not get more complex in a comprehension
newTeam={k: v for team in (team1, team2) for k, v in team.items()}
newTeam

{'Jones': 24,
 'Jameson': 18,
 'Smith': 58,
 'Burns': 7,
 'White': 12,
 'Macke': 88,
 'Pierce': 4}

## set comprehension

In [16]:
ctemps = [0, 10, 12, 14, 10, 23, 12, 34, 34, 100, 100]
ftemps1 = [(t*9/5) for t in ctemps]
# when we do not use kev-value pairs
# curly brackets create a set, sets do not allow duplicates
ftemps2 = {(t*9/5) for t in ctemps}
print(ftemps1)
print(ftemps2)

[0.0, 18.0, 21.6, 25.2, 18.0, 41.4, 21.6, 61.2, 61.2, 180.0, 180.0]
{0.0, 41.4, 18.0, 180.0, 21.6, 25.2, 61.2}


In [21]:
sTemp = "The quick brown fox jumped over the lazy dog"
chars = {c.upper() for c in sTemp if not c.isspace()}
print(sorted(chars))

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']


# Iterators
- help cleaning up the code 
- uses them instead of for-loops

## iter

In [31]:
days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
i=iter(days)
print(next(i))
print(next(i))
print(next(i))
print(next(i))
print(next(i))

Sunday
Monday
Tuesday
Wednesday
Thursday


## enumerate

In [28]:
for i, m in enumerate(days, start=1):
    print(i, m)

1 Sun
2 Mon
3 Tue
4 Wed
5 Thu
6 Fri
7 Sat


## zip

In [32]:
days = [ 'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' ]
day_ger = ['Sonntag', 'Montag', 'Dienstag', 'Mittwoch', 'Donerstag', 'Freitag', 'Samstag']
for m in zip(days, day_ger):
    print(m)

('Sunday', 'Sonntag')
('Monday', 'Montag')
('Tuesday', 'Dienstag')
('Wednesday', 'Mittwoch')
('Thursday', 'Donerstag')
('Friday', 'Freitag')
('Saturday', 'Samstag')


In [ ]:
l = list(zip([1, 2, 3], ['a', 'b', 'c']))   # pairs the two list to a list of tuples
d = dict(zip([1, 2, 3], ['a', 'b', 'c']))   # pairs two lists to a dict
print(l)
print(d)


[(1, 'a'), (2, 'b'), (3, 'c')]
{1: 'a', 2: 'b', 3: 'c'}


In [34]:
for i, m in enumerate(zip(days, day_ger)):
    print(i, m[0], '=', m[1], 'in german')

0 Sunday = Sonntag in german
1 Monday = Montag in german
2 Tuesday = Dienstag in german
3 Wednesday = Mittwoch in german
4 Thursday = Donerstag in german
5 Friday = Freitag in german
6 Saturday = Samstag in german


# itertools

In [7]:
import itertools

In [9]:
# cycle iterator cycles over a collection
seq1 = ['Joe', 'Jana', 'Joseph']
cycle1 = itertools.cycle(seq1)
print(next(cycle1))
print(next(cycle1))
print(next(cycle1))
print(next(cycle1)) # cycles to the beginning

Joe
Jana
Joseph
Joe


In [10]:
# count iterator
count1 = itertools.count(100, 10)
print(next(count1))
print(next(count1))
print(next(count1))
print(next(count1))
print(next(count1))

100
110
120
130
140


In [13]:
# accumulate iterator - running addition
vals = [10, 20, 60, 40, 50, 15, 30]
accu = itertools.accumulate(vals)
print(list(accu))

[10, 30, 90, 130, 180, 195, 225]


In [14]:
# goes over the numbers and sticks with the max 
accu2 = itertools.accumulate(vals, max)
print(list(accu2))

[10, 20, 60, 60, 60, 60, 60]


In [16]:
# chain - chains to sequences
x =itertools.chain('ABCD', '1234')
print(list(x))

['A', 'B', 'C', 'D', '1', '2', '3', '4']


In [23]:
def fct(x):
    return x < 40

In [25]:
print(vals)
# drops values as long as (fct retuns True) trigger point is not reached
print(list(itertools.dropwhile(fct, vals)))
# returns values until (fct is False) trigger is reached
print(list(itertools.takewhile(fct, vals)))

[10, 20, 60, 40, 50, 15, 30]
[60, 40, 50, 15, 30]
[10, 20]


# advanced collections
- found in  collections module
- The collections module provides alternatives to built-in container data types
- namedtuple: tuple with named fields
- OrderedDict, defaultdict: dictionaries with special properties
- Counter: counts distinct values
- deque: double-ended list

__namedtuple()__

The namedtuple() function returns a tuple-like object with named fields. </br>
These field attributes are accessible by lookup as well as by index. </br>
Named tuples allow you to create tuples and assign meaningful names to the positions of the tuple’s elements.</br>
Technically, a named tuple is a subclass of tuple. On top of that, it adds property names to the positional elements.</br>

In [37]:
from collections import namedtuple
student = namedtuple('student', ["name", "age", "marks"])  # declares a student class having name, age and marks as fields
s1 = student("Imran", 21, 98)  #create a new object of this namedtuple
print(s1.name)  # fields are accessible by attribute lookup
print(s1[2]) # or index

Imran
98


In [40]:
Point = namedtuple('Point', 'x y') # alternative naming 
p1 = Point(10, 20)
p2 = Point(30, 40)

print(p1)
print(p2.x) # instead of p2[1]
print(p1.y)

p1 = p1._replace(x=15)  # change value using it's name
print(p1)

Point(x=10, y=20)
30
20
Point(x=15, y=20)


## OrderedDict()
- The OrderedDict() function is similar to a normal dictionary object in Python. 
- However, it remembers the order of the keys in which they were first inserted. 
- Upon traversing the dictionary, pairs will appear in the order of their insertion. 
- pop item from the top is possible

In [80]:
from collections import OrderedDict 
d1 = OrderedDict()
d1['A'] = 65
d1['C'] = 67
d1['B'] = 66
d1['D'] = 68

for k,v in d1.items():
    print (k,v)

A 65
C 67
B 66
D 68


In [81]:
first = d1.popitem(False) # removes first item in the dict
last = d1.popitem() # removes the last
print(first)
print(last)
print(d1)

('A', 65)
('D', 68)
OrderedDict([('C', 67), ('B', 66)])


In [82]:
a = OrderedDict({'a': 1, 'b': 2, 'c': 3})
b = OrderedDict({'a': 1, 'c': 3, 'b': 2})
c = {'a': 1, 'c': 3, 'b': 2}

print(a==b) # order matters for OrderdDict
print(a==c) # for ordinary dict order does not matter

False
True


## __defaultdict__

In [ ]:
fruits = ['apple', 'pear', 'banana', 'apple', 'peach', 'cherry', 'banana']
fruitCounter={} # dict that should count the fruits

for fruit in fruits:
    if fruit in fruitCounter.keys():   # error if there is no fruit initilized in the dict
        fruitCounter[fruit] += 1
    else:
        fruitCounter[fruit] = 1  # thus we need to check and insert a key (fruit) first 

fruitCounter

{'apple': 2, 'pear': 1, 'banana': 2, 'peach': 1, 'cherry': 1}

In [49]:
# above code can be shortened and the
# initilizing switched with a defaultdict
from collections import defaultdict
# int is the factory fct. and produces a default key if there is none
fruitCounter = defaultdict(lambda: 100)

for fruit in fruits:
# now no chceking for key is required bc. defaultdict sets a default key
        fruitCounter[fruit] += 1

for k, v in fruitCounter.items():
    print(k,':', v)

apple : 102
pear : 101
banana : 102
peach : 101
cherry : 101


## Counter
- dict that counts hashable objects

In [69]:
from collections import Counter
class1 = ['Bob', 'Jenny', 'Hannes', 'Greg', 'Boris', 'Lara', 'Hannah', 'Suse', 'Kara', 'Bob']
class2=['Lara', 'Steffen', 'Simon', 'Becky', 'Lila', 'Gunna', 'Lena', 'Rose', 'Greg', 'Jenny']

c1= Counter(class1)
c2= Counter(class2)

In [63]:
print(c1['Bob']) # how many Bob's are in c1?
print(sum(c1.values()), ' students in class 1')

2
10  students in class 1


In [64]:
c1.update(class2) # combine two sets
print(sum(c1.values()), ' students in class 1')

20  students in class 1


In [65]:
print(c1.most_common(3))

[('Bob', 2), ('Jenny', 2), ('Greg', 2)]


In [67]:
c1.subtract(class2) # separate the sets again
print(c1.most_common(3))

[('Bob', 2), ('Hannes', 1), ('Boris', 1)]


In [70]:
print(c1 & c2) # common objects in both

Counter({'Jenny': 1, 'Greg': 1, 'Lara': 1})


## deque
- double-ended queue, pronounced 'deck'
- accessible from both sides
- appendleft(), append(), popleft(), pop()
- rotate(): 

In [89]:
string.ascii_lowercase

'abcdefghijklmnopqrstuvwxyz'

In [106]:
from collections import deque
import string
d = deque(string.ascii_lowercase) # initilized with lowercase letters

# lenght of string.ascii_lowercas
print("Item count: ", str(len(d)))

# convert lower to uppercase letters
for elem in d:
    print(elem.upper(), end=',')

Item count:  26
A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z,

In [102]:
d.pop()
d.popleft()
d.appendleft(1)
d.append(30)
print(d)

deque([1, 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 30])


In [107]:
print(d)
d.rotate(10)
print(d)

deque(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'])
deque(['q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p'])


# generator functions
- generator functions return a lazy iterator an iterator object with a sequence of values
- These are objects that you can loop over but unlike lists, lazy iterators do not store their contents in memory. 
- Using __yield__ will result in a generator object. - Using __return__ will result in the first line of the file only.
-  Calling a generator function creates an generator object.  However, it does not start running the function.
- The function only executes on next()
-  The difference between yield and return is that yield returns a value and pauses the execution while maintaining the internal states, </br>
whereas the return statement returns a value and terminates the execution of the function. 
-  The generator is called just like a normal function. However, __its execution is paused on encountering the yield keyword.__ </br>
This sends the first value of the iterator stream to the calling environment. However, __local variables and their states are saved internally.__ </br>
This includes any variable bindings local to the generator, the instruction pointer, the internal stack, and any exception handling.</br>
- This allows you to resume function execution whenever you call one of the generator’s methods.  </br>
- That way, when next() is called on a generator object (either explicitly or implicitly within a for loop), </br>
the previously yielded variable num is incremented, and then yielded again. 
- Unless your generator is infinite, __you can iterate through it one time only.__
- Once all values have been evaluated, the generator is deemed exhausted. The iteration will stop and the for loop will exit. 
- If you used next(), then instead you’ll get an explicit StopIteration exception.

One of the __advantages__ of the generator over the iterator is that __elements are generated dynamically.__</br>
Since the next item is generated only after the first is consumed, it is __more memory efficient__ than the iterator. 

    1. Do you need the entire results in memory?
    2. Do you need to reuse the raw results as is?
    3. Is your result reasonably small to fit in the memory?
    4. Do you want to process the results after you have obtained all the results?

If all of the above is yes, then an iterator should suffice. Otherwise, you may want to consider using a generator to benefit from the delayed execution and yielding on the fly.


In [375]:
def get_sequence_upto(x):
    for i in range(x):
        yield i

In [379]:
seq = get_sequence_upto(5)
print(next(seq))
print(next(seq))
print(next(seq))
print(next(seq))
print(next(seq))
print(next(seq))  # The function finally terminates when next() encounters the StopIteration error.

0
1
2
3
4


StopIteration: 

In [383]:
# In the following example, function square_of_sequence() acts as a generator.
# It yields the square of a number successively on every call of next().

def square_of_sequence(x):
    for i in range(x):
        yield i * i


gen = square_of_sequence(5)

while True:
    try:
        print("Received on next(): ", next(gen))
    except StopIteration:
        break

Received on next():  0
Received on next():  1
Received on next():  4
Received on next():  9
Received on next():  16


In [384]:
# We can use the for loop to traverse the elements over the generator. 
# In this case, the next() function is called implicitly and the StopIteration is also automatically taken care of.
squres = square_of_sequence(5)
for sqr in squres:
    print(sqr)

0
1
4
9
16


## generator comprehension
- shorter way of defining simple generator functions.
- They’re useful in the same cases where list comprehensions are used, with an added benefit: </br>
you can create them without building and holding the entire object in memory before iteration. 

In [397]:
liste_ = [x * x for x in range(10) if x %2 ==0]
print(liste_)

# (expression for i in s if condition)
gen = (x * x for x in range(10) if x % 2 == 0)
print(gen)
print(next(gen))
print(next(gen))
print(next(gen))

[0, 4, 16, 36, 64]
<generator object <genexpr> at 0x7ff81ec957b0>
0
4
16


In [395]:
# The generator expression can also be passed in a function. It should be passed without parentheses.
sum(x * x for x in range(10))


285

## A Generator Solution

In [408]:
with open("../data/test.txt") as wwwlog:
    bytecolumn = (line.rsplit(" ", 1)[1] for line in wwwlog)
    bytes_sent = (int(x) for x in bytecolumn if x != '-')
    print("Total", sum(bytes_sent))

Total 135667


## Performance - of generator objects
- list you get from the list comprehension is 87,624 bytes, while the generator object is only 120. 
- This means that the list is over 700 times larger than the generator object!

In [249]:
import sys
nums_squared_lc = [i * 2 for i in range(10000)]
print(sys.getsizeof(nums_squared_lc))

nums_squared_gc = (i ** 2 for i in range(10000))
print(sys.getsizeof(nums_squared_gc))

87616
112


## memory vs. speed
- If the list is smaller than the running machine’s available memory, then list comprehensions can be faster to evaluate than the equivalent generator expression.
- Here, you can see that summing across all values in the list comprehension took about a third of the time as summing across the generator. 
- If speed is an issue and memory isn’t, then a list comprehension is likely a better tool for the job.

In [252]:
import cProfile
cProfile.run('sum([i * 2 for i in range(10000)])')

         5 function calls in 0.001 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    0.001    0.001 <string>:1(<listcomp>)
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.sum}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [253]:
cProfile.run('sum((i * 2 for i in range(10000)))')

         10005 function calls in 0.002 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    10001    0.001    0.000    0.001    0.000 <string>:1(<genexpr>)
        1    0.000    0.000    0.002    0.002 <string>:1(<module>)
        1    0.000    0.000    0.002    0.002 {built-in method builtins.exec}
        1    0.001    0.001    0.002    0.002 {built-in method builtins.sum}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




## send, throw, close - Advanced Generator Methods

### send()
- Resumes the execution and “sends” a value into the generator function.

In [299]:
def f(x=None):
    while True:
        x = yield
        yield x*2
             
g = f()
next(g)
g.send(4)

8

In [300]:
next(g)
g.send(10)

20

### throw
-allows you to throw exceptions with the generator.

### close
- allows you to stop a generator.

# Classes and Objects
- Almost everything in Python is an object, with its properties and methods.
- Object oriented programming OOP is about data (state) and behavior
- A __Class__ is like an object constructor, or a "blueprint" for creating objects.
- a class gives the shape to an object and contains the default data/instance __attribute__
- __objects__ are an Instance of a class - an instance is an object that is built from a class and contains real data. 
- the __state__ of an object is stored in instance attributes 
- __behavior__ is define by instance functions in the class definition
- behavoir:  set of methods
- Put another way, a class is like a form or questionnaire. An instance is like a form that has been filled out with information.

In [15]:
class Point:                                        # Uppercase for classes Pascal-naming-convention
    def __init__(self, x=0, y=0):       # initialise constructor with default values
        self.x = x                                    # self is a reference to the current obeject and self.x = x gives a value to the attribute self.x
        self.y = y                                    #  

    # instance methods - Instance methods are functions that are defined inside a class 
    # and can only be called from an instance of that class. Just like .__init__(), an instance method’s first parameter is always self.
    def mov(self):
        print("move")

    def draw(self):
        print("draw")

- object - is an instance of a class
- one of the biggest advantages of using classes to organize data <br>
 is that instances are guaranteed to have the attributes you expect. 

In [56]:
point1 = Point()        # instanciate an object of a Class - specific object of type Point
point1.x = 10            # attributes of a particular object
point1.y = 20            # y attribute

print(point1.x)         # we can access the properties of that object.
point1.draw()           # method we defined for the class

point2 = Point()        # second object
print(point2.x)         # there is no attribute x

point3 = Point(10, 20)  # 10 is passed to the x attribute 20 to the y attribute
print(point3.x, point3.y)

10
draw
0
10 20


## the constructor - def \_\_init\_\_(self, x, y): 
- Every time anobject is created, .__init__() sets the initial state of the object by assigning the values of the object’s properties. 
- That is, .__\_\_init\_\_()__ initializes each new instance of the class.


- When a new class instance is created, the instance is automatically passed to the self parameter in .__init__() so that new attributes can be defined on the object.
- The self parameter is a reference to the current instance of the class, </br> 
and is used to access variables that belong to the class. **It does not have to be named self, </br> 
you can call it whatever you like**, but it has to be the first parameter of any function in the class:

- Attributes created in .__init__() are called __instance attributes__. An instance attribute’s value is specific to a particular instance of the class. 
- On the other hand, __class attributes__ are attributes that have the same value for all class instances. You can define a class attribute by assigning a value to a variable name outside of .__init__().


In [21]:
class Person:
    species = "human"  # class attribute
    
    # here 'self' is named person, the self parameter allows us to access the attributes/ methods of a class
    # self (or whatever name you give it) references to the current object
    def __init__(person, name="no_name", age="no_age"):
        person.name = name    # self (here 'person') is placeholder first for the class name then for the object name
        person.age = age            # 
        
    # this  is a class method which every object of that class inherits
    # 'instance' is also just a placeholder for the instance name
    def talk(instance): 
        print(f"Hi it's me {instance.name}, i'm {instance.age} years old")

# Creating a new object from a class is called instantiating an object. Instances are unique.
p1 = Person("John", 36)  # p1 is an instance of the class Person, self in class becomes p1
print(p1.name)
p1.talk()                              # the instance p1 inerited the class method 'talk'

p2 = Person()
print(p2.name, "/", p2.age)
p2.name ="Sascha"
p2.age = "17"  # modify object properties
print(p2.age)
p2.talk()

John
Hi it's me John, i'm 36 years old
no_name / no_age
17
Hi it's me Sascha, i'm 17 years old


In [22]:
del p2.name  # delete properties
#del p2  # delete object
print(p2.age)
p2.myfct2()


17


AttributeError: 'Person' object has no attribute 'myfct2'

## setattr, hasattr, getattr, delattr

In [23]:
print(hasattr(p1, "profession"))
print(hasattr(p1, "name"))


False
True


In [24]:
setattr(Person, "health", "good")  # set a class attibute with a default
print(hasattr(Person, "health"))


True


In [26]:
print(getattr(p1, "health"))
print(getattr(p1, "name"))

good
John


In [27]:
delattr(Person, "health")
print(hasattr(Person, "health"))

False


In [28]:
print(p1)

## inhertance
-  inheritance - reuse code, define once (dry)
- Instances of child classes inherit all of the attributes and methods of the parent class:
- Child classes can override or extend the attributes and methods of parent classes. 
- In other words, child classes inherit all of the parent’s attributes and methods but can also specify attributes and methods that are unique to themselves.

In [ ]:
# parent class with attributes and methods
class Human():
    def __init__(self, name="nobody", age="30 something", gender="open-mided"):
        self.name=name
        self.age=age
        self.gender=gender

    def talk(self):
        print(f"I'm {self.name}, {self.age} years old, and {self.gender}")

human = Human("Jesus", 24, "male")
human.gender
human.talk()


I'm Jesus, 24 years old, and male


In [ ]:
# child class - inherits from parent class (Human)
class Worker(Human):

    def __init__(self, name, age, gender, profession, experience): # all attributes of Worker
        super().__init__(name, age, gender) # attributes of parent class, which the child inherits
        self.profession = profession        # add specific Worker attributes
        self.experience = experience

    # specfific Worker method
    def skills(self):
        print(f"{self.name}, is a {self.profession} and has experience as {self.experience}")


worker = Worker('Igor', 34, 'divers', 'construction worker', 'bootlegger')
worker.talk()
worker.skills()

I'm Igor, 34 years old, and divers
Igor, is a construction worker and has experience as bootlegger


In [156]:
# first create parent class
class Mammal:
    def walk(self):
        print("walk")


# create a child class -  that inherits the functionality from another class, by sending
# the parent class as a parameter when creating the child class:
class Dog(Mammal):      # Dog inherits the Mammal functions
    def bark(self):
        print("bark")


# secon child class without extra functionality
class Cat(Mammal):
    pass                # pass does nothing


# Use the Mammal class to create object, and then execute the methods:
mam1 = Mammal();    
print(type(mam1));  
mam1.walk();     
print("\n")

dog1 = Dog();   
dog1.walk();    
dog1.bark();     
print(type(dog1));     
print("\n")

cat1 = Cat()
cat1.walk()
print(type(cat1))
print(isinstance(cat1, Dog))
print(isinstance(cat1, Cat))


<class '__main__.Mammal'>
walk


walk
bark
<class '__main__.Dog'>


walk
<class '__main__.Cat'>
False
True


### Extend the Functionality of a Parent Class
- To override a method defined on the parent class, you define a method with the same name on the child class. 
- changes to the parent class automatically propagate to child classes, as long as the attribute or method being changed isn’t overridden in the child class.

In [170]:
# parent class
class Mammal:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def walk(self):
        print("walk")

    def speak(self, sound):
        return f"{self.name} says {sound}"

# child class
class Dog(Mammal):
    # override parent method
    def speak(self, sound="Woof"):
        return f"{self.name} barks {sound}"

class Cat(Mammal):

    def speak(self, sound="Mau"):
        return f"{self.name} sings {sound}"


dog = Dog("Bernd", 7)
print(dog.speak())

cat =Cat("Muschi", 3)
print(cat.speak())
# you can still call .speak() with a different sound:
print(cat.speak("Fauch"))

Bernd barks Woof
Muschi sings Mau
Muschi sings Fauch


### super()
- we may want to override a parent method in a child class, but we also dont't want to lose the changes made to the parent method
- thus we can call the aprent method in the child class with super
- super() does much more than just search the parent class for a method or an attribute. </br>
It traverses the entire class hierarchy for a matching method or attribute. If you aren’t careful, super() can have surprising results.

In [173]:
class Wolf(Mammal):

    def speak(self, sound="howl"):
        return super().speak(sound)

wolf = Wolf("Wolfgang", 3)
wolf.speak()


'Wolfgang says howl'

#  Classes advanced

## enumerations
-  an enumeration is a set of members that have associated unique constant values. Enumeration is often called enum.
- Enumerations are immutable. It means you cannot add or remove members once an enumeration is defined. And you also cannot change the member values.
- Python provides you with the enum module that contains the Enum type for defining new enumerations. 
- And you define a new enumeration type by subclassing the Enum class.
- enumeration’s members are constants. 
- Therefore, their names are in uppercase letters by convention.
- Enumeration members are always hashable. It means that you can use the enumeration members as keys in a dictionary or as elements of a Set.
- The hash() function accepts an object and returns the hash value as an integer. When you pass an object to the hash() function, Python will execute the __hash__ special method of the object.

In [13]:
from enum import Enum, unique, auto

@ unique # decorater that prohibits duplicate values
class Color(Enum):  # Color class that inherits from the Enum type
    # members of the Color enumeration
    # names have to be unique
    RED = 1
    GREEN = 2
    BLUE = 3
    # BLUE = 5 # is not allowed - bc its a duplicate
    YELLOW = auto() # assigns a random value

In [19]:
print(Color(3))  # numeration is callable, you can get a member by its value.
print(type(Color.RED)) # type
print(isinstance(Color.RED, Color)) # Color.Red is an instance of the Color enumeration
print(Color.RED.name, Color.RED.value) # name and value
print(repr(Color.YELLOW))  # string representation of the object.

Color.BLUE
<enum 'Color'>
True
RED 1
<Color.YELLOW: 4>


In [17]:
# enums (as hash value) can be keys in a dictionary
rgb = {
    Color.RED: '#ff0000', 
    Color.GREEN: '#00ff00', 
    Color.BLUE: '#0000ff'
    }
    
print(rgb[Color.RED])

#ff0000


In [20]:
for color in Color:
    print(color)

Color.RED
Color.GREEN
Color.BLUE
Color.YELLOW


## Special \_\_dunder\_\_ methods
- Dunder (double underscore) Methods , Special Methods and Magic methods are the same thing in Python.
- Dunder Methods makes our class compatible with inbuilt functions like abs(), len(), str() etc. <br>
and extend the functionality of the class.
- They give the ability to create classes that behave like native data structures like lists, tuples, dictionary, set etc.
- Users don’t need to remember each and every method to do a certain task, <br>
just use inbuilt function and pass the object with required parameters.
- The methods have a predefined syntax which is available to implement in our class.
- there are many dunder methodes: https://holycoders.com/python-dunder-special-methods/

### class string functions
- __\_\_str\_\_(self)__ for str() function. </br> 
Return a string to print the object. Intended for users to see a pretty and useful output. If not implemented, __repr__ will be used as a fallback.
- by overriting the function below you can control how your object is represented in string and bytes format
- **object.__str__(self)**<br> 
Return a string to print the object. Intended for users to see a pretty and useful output.
- **object.__repr__(self)**<br>
It returns a string to print the object. Intended for developers to debug. Must be implemented in any class.
- **object.__format__(self, format_spec)**<br>
Evaluate formatted string literals like % for percentage format and ‘b’ for binary.
- **object.__bytes__(self)** <br>
Return a byte object which is the byte string representation of the object.

In [46]:
class Person():
    def __init__(self):
        self.fname='Marco'
        self.lname='Duno'
        self.age= 25
    def __repr__(self): # detailed for debugging
        return"<Person Class - fname:{0}, lname:{1}, age:{2}".format(
            self.fname, self.lname, self.age
            )

    def __str__(self) -> str: # string for info about object
        return"Person ({0} {1} is {2})".format(self.fname, self.lname, self.age)

    def __bytes__(self):
        # string 
        val = "Person: {0}:{1}:{2}".format(self.fname, self.lname, self.age)
        return bytes(val.encode('utf-8'))


In [47]:
cls1=Person()
# repr and str return non-human-readable representations
# the definition of reor and str give the object a readable string output
'''
<__main__.Person object at 0x7f529b5894f0>
<__main__.Person object at 0x7f529b5894f0>
Formatted: <__main__.Person object at 0x7f529b5894f0>
'''
print(repr(cls1))
print(str(cls1))
print("Formatted: {0}".format(cls1))
print(bytes(cls1))


<Person Class - fname:Marco, lname:Duno, age:25
Person (Marco Duno is 25)
Formatted: Person (Marco Duno is 25)
b'Person: Marco:Duno:25'


### class attribute functions
- **object.__getattr__(self, attr)**<br>
“catch” references to attributes that don’t exist in your object.<br>
using the __getattr__ magic method, we can intercept that inexistent attribute lookup and do something so it doesn’t fail
- object.__getattribute__(self, attr)<br>
\_\_getattribute\_\_ is similar to \_\_getattr\_\_, with the important difference that \_\_getattribute\_\_ will intercept EVERY attribute lookup, doesn’t matter if the attribute exists or not.<br>
- object.__setattr__(self, attr)
- object.__delattr__(self, attr)
- object.__dir__(self, attr)

In [64]:
class myColor():
    def __init__(self):
        self.red=50
        self.green=75
        self.blue=100

    # dynamically return a value
    def __getattr__(self, attr):
        if attr == "rgbcolor":
            return (self.red, self.green, self.blue)
        elif attr=='hexcolor':
            return "#{0:02x}{1:02x}{2:02x}".format(self.red, self.green, self.blue)
        else:
            raise AttributeError  # indicating an atrribute we don't know was requested 

    # give the user of the class the ability to set an attribute value
    def __setattr__(self, attr, val):
        if attr == 'rgbcolor':
            self.red = val[0]
            self.green = val[1]
            self.blue = val[2]
        else: # always put the super()__setattr__ as a default 
            super().__setattr__(attr, val)

    # list the available properties
    def __dir__(self):
        return('red', 'green', 'blue', 'rgbcolor', 'hexcolor')


In [66]:
cls1=myColor()
print(cls1.rgbcolor)
print(cls1.hexcolor)
cls1.rgbcolor = (125, 200, 86)
print(cls1.rgbcolor)
print(cls1.hexcolor)
print(cls1.red)
print(dir(cls1))

(50, 75, 100)
#324b64
(125, 200, 86)
#7dc856
125
['blue', 'green', 'hexcolor', 'red', 'rgbcolor']


### class numeric operators
Arithmetic Operators
- **object.\_\_add\_\_(self, anotherObj)**: self + other
- **object.\_\_sub\_\_(self, anotherObj)** 
-  **object.\_\_mul\_\_(self, anotherObj)**  
- **object.\_\_matmul\_\_(self, anotherObj)** 
- **object.\_\_truediv\_\_(self, anotherObj)**  
- **object.\_\_floordiv\_\_(self, anotherObj)**  
- **object.\_\_pow\_\_(self, anotherObj)**
- **object.\_\_and\_\_(self, anotherObj)**
- **object.\_\_or\_\_(self, anotherObj)**

In [79]:
class Point():
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __repr__(self):
        return"<Point x:{0}, y.{1}>".format(self.x, self.y)

    def __sub__(self, other):
        return Point(self.x - other.x, self.y - other.y)

    # inplace addition
    def __iadd__(self, other):
        self.x += other.x
        self.y += other.y
        return self


In [80]:
p1 = Point(10, 20)
p2 = Point(15, 5)

In [83]:
print(p1)
print(p2)
print(p1+p2)
print(p1-p2)

<Point x:10, y.20>
<Point x:15, y.5>
<Point x:25, y.25>
<Point x:-5, y.15>


### class comparison operators
- **object.\_\_lt\__\(self, other)**
- **object.\_\_le\_\_(self, other)**
- **object.\_\_eq\_\_(self, other)**
- **object.\_\_ne\_\_(self, other)**
- **object.\_\_ge\_\_(self, other)**
- **object.\_\_gt\_\_(self, other)**

In [94]:
class Employee():

    def __init__(self, fname, lname, level, yrsService):
        self.fname = fname
        self.lname = lname
        self.level = level
        self.seniority = yrsService

    def __ge__(self, other):
        if (self.level == other.level):
            return self.seniority >= other.seniority
        return self.level >= other.level

    def __gt__(self, other):
        if (self.level ==other.level):
            return self.seniority > other.seniority
        return self.level > other.level

    def __lt__(self, other):
        if (self.level ==other.level):
            return self.seniority < other.seniority
        return self.level < other.level

    def __le__(self, other):
        if (self.level == other.level):
            return self.seniority <= other.seniority
        return self.level <= other.level

    def __eq__(self, other):
        return self.level == other.level

In [115]:
dept =[]
dept.append(Employee('Tim', 'Sims', 3, 12))
dept.append(Employee('John', 'Doe', 3, 9))
dept.append(Employee('Jane', 'Smith', 6, 6))
dept.append(Employee('Rebecca', 'Robinson', 5, 11))
dept.append(Employee('Tyler', 'Durden', 5, 12))

In [113]:
print(dept[0] > dept[1])
print(dept[3] < dept[4])

True
False


In [116]:
emps = sorted(dept) # sorts by level and seniority

for emp in emps:
    print(emp.lname)

Doe
Sims
Robinson
Durden
Smith


# logging
- captures and records events while app is running
- useful debugging feature
- real-time debugging is not always possible
- events can be categorized for easier analysis
- provides a tranaction record of  a program's execution
- highly customizable output<br><br>
- **logging.debug('This is a debug message')**: diagnostic info for debugging
- **logging.info('This is an info message')**: general info about execution results
- **logging.warning('This is a warning message')**: sth. unexpected or approaching problem
- **logging.error('This is an error message')**: unable to perform specific operation due to problem
- **logging.critical('This is a critical message')**: prog may not be able to continue, serious error

## Customizations
- **%(acstime)s**: human-readable date format when the log was created
- **%(filename)s**: filename where the log message originated
- **%(funcName) s**: unction name where the log message originated
- **%(levelname)s**: string representation of the message level (DEBUG, INFO, etc) 
- **%(levelno)d**: numeric representation of the message level
- **%(lineno)d**: source line number where the logging call was issued (if available)
- **%(message)s**: the logged meaasge string itself
-**%(module)s**: module name portion of the filename where the message was logged

In [127]:
import logging
logging.debug('This is a debug message')
logging.info('This is an info message')
logging.warning('This is a warning message')
logging.error('This is an error message')
logging.critical('This is a critical message')

ERROR:root:This is an error message
CRITICAL:root:This is a critical message


In [ ]:
# Notice that the debug() and info() messages didn’t get logged.
# This is because, by default, the logging module logs the messages
# with a severity level of WARNING or above. You can change that
# by configuring the logging module to log events of all levels if you want.
# you need to restart the kernel bc. the basicConfig can only be configured once

In [1]:
import logging

extData={'user': 'joe@example.com'}

def anotherFunction():
    logging.debug("A debug-level message", extra=extData)

# datetime, process ID along with the level,
# fct name, line number and message
fmtstr = "User: %(user)s %(asctime)s: %(process)d: %(levelname)s: \
                 %(funcName)s Line:%(lineno)d %(message)s"

# customized date format
datestr = "%m/%d/%Y %I:%M:%S %p"

logging.basicConfig(
    level=logging.DEBUG, # lower logging level specified
    filemode='w', # 'write log into file default is 'a' for apppend
    filename='output.log',
    format=fmtstr,
    datefmt=datestr)

logging.info("This is an info-level log message", extra=extData)
logging.info("This is will get logged to a file", extra=extData)
logging.debug("This is a debug log message", extra=extData)
logging.warning('This a warning message', extra=extData)
anotherFunction()

# modules
- Python files are called modules and they are identified by the .py file extension. 
- A module can define functions, classes, and variables.
- Now  use the module, by import the module in your current file
- You can create an alias when you import a module, by using the __as__ keyword

In [310]:
import convert_measure as c         # refer to the entire module & use a shorthand
print(c.inch_cm(10))

25.4


In [ ]:
from convert_measure import inch_cm # pick a function of a module
print(inch_cm(10))

## dir()
- ‘dir()’ can be used to find out which names a module defines. It returns a sorted list of strings.

In [432]:
print(dir(time))

['CLOCK_BOOTTIME', 'CLOCK_MONOTONIC', 'CLOCK_MONOTONIC_RAW', 'CLOCK_PROCESS_CPUTIME_ID', 'CLOCK_REALTIME', 'CLOCK_THREAD_CPUTIME_ID', '_STRUCT_TM_ITEMS', '__doc__', '__loader__', '__name__', '__package__', '__spec__', 'altzone', 'asctime', 'clock_getres', 'clock_gettime', 'clock_gettime_ns', 'clock_settime', 'clock_settime_ns', 'ctime', 'daylight', 'get_clock_info', 'gmtime', 'localtime', 'mktime', 'monotonic', 'monotonic_ns', 'perf_counter', 'perf_counter_ns', 'process_time', 'process_time_ns', 'pthread_getcpuclockid', 'sleep', 'strftime', 'strptime', 'struct_time', 'thread_time', 'thread_time_ns', 'time', 'time_ns', 'timezone', 'tzname', 'tzset']


 ## .\_\_doc\_\_
 - You can see the internal documentation (if available) of a module name by looking at  .\_\_doc\_\_.

In [458]:
import time

print(time.strftime.__doc__)

strftime(format[, tuple]) -> string

Convert a time tuple to a string according to a format specification.
See the library reference manual for formatting codes. When the time tuple
is not present, current time as returned by localtime() is used.

Commonly used format codes:

%Y  Year with century as a decimal number.
%m  Month as a decimal number [01,12].
%d  Day of the month as a decimal number [01,31].
%H  Hour (24-hour clock) as a decimal number [00,23].
%M  Minute as a decimal number [00,59].
%S  Second as a decimal number [00,61].
%z  Time zone offset from UTC.
%a  Locale's abbreviated weekday name.
%A  Locale's full weekday name.
%b  Locale's abbreviated month name.
%B  Locale's full month name.
%c  Locale's appropriate date and time representation.
%I  Hour (12-hour clock) as a decimal number [01,12].
%p  Locale's equivalent of either AM or PM.

Other codes may be available on your platform.  See documentation for
the C library strftime function.



## built-in modules

In [ ]:
#  list of all available modulesb
help("modules")

### os module
- provides functions for creating and removing a directory (folder), fetching its contents, changing and identifying the current directory, etc. 

In [321]:
import os
os.getcwd()
os.mkdir("test")  # create folder
os.chdir("C:\MyPythonProject")  # changing current workign directory
os.chdir("..")  # to set the current directory to the parent directory use ".."
os.rmdir("test")  # remove dirb
os.listdir()

'/home/mz/code/Stuff/pyfiles'

### sys module
- provides functions and variables used to manipulate different parts of the Python runtime environment.

In [488]:
sys.platform

'linux'

In [442]:
import sys
print("You entered: ", sys.argv[1], sys.argv[2], sys.argv[3])


You entered:  --ip=127.0.0.1 --stdin=9038 --control=9036


__sys.argv__
- Python lets you access whatever you have passed to a script while calling it. The ‘command line’ content is stored in the sys.argv list.
- The item at index 0 in this list is always the name of the script. The rest of the arguments are stored at the subsequent indices. 

when you create a file with the content:
___
import sys</br>
print("You entered: ",sys.argv[1], sys.argv[2], sys.argv[3])
___
and execute from the commandline as follows:</br>
___
$> python test.py Python C# Java
>>You entered: Python C# Java 

- Above, sys.argv[1] contains the first argument 'Python', 
- sys.argv[2] contains the second argument 'C#', 
- and sys.argv[3] contains the third argument 'Java'. 
- sys.argv[0] contains the script file name test.py. 


__sys.exit__
- This causes the script to exit back to either the Python console or the command prompt. 
- This is generally used to safely exit from the program in case of generation of an exception. 

__sys.maxsize__

Returns the largest integer a variable can take. </br>
maxsize attribute of the sys module fetches the largest value a variable of data type Py_ssize_t can store. </br>
It is the Python platform’s pointer that dictates the maximum size of lists and strings in Python. </br>
The size value returned by maxsize depends on the platform architecture:</br>

    32-bit: the value will be 2^31 – 1, i.e. 2147483647</br>
    64-bit: the value will be 2^63 – 1, i.e. 9223372036854775807

In [345]:
# importing the module
import sys
max_val = sys.maxsize       # fetching the maximum value
list = range(max_val)         # creating list with max size
print(max_val)
print(len(list))                     # displaying the length of the list

9223372036854775807
9223372036854775807


__sys.path__

This is an environment variable that is a search path for all Python modules. 

In [348]:
sys.path

['/home/mz/code/Stuff/pyfiles',
 '/home/mz/.pyenv/versions/3.8.12/envs/lewagon/lib/python3.8/site-packages/git/ext/gitdb',
 '/home/mz/code/MaCoZu/data-challenges/04-Decision-Science',
 '/home/mz/code/Stuff/pyfiles',
 '/home/mz/.pyenv/versions/3.8.12/lib/python38.zip',
 '/home/mz/.pyenv/versions/3.8.12/lib/python3.8',
 '/home/mz/.pyenv/versions/3.8.12/lib/python3.8/lib-dynload',
 '',
 '/home/mz/.pyenv/versions/3.8.12/envs/lewagon/lib/python3.8/site-packages',
 '/home/mz/.pyenv/versions/3.8.12/envs/lewagon/lib/python3.8/site-packages/IPython/extensions',
 '/home/mz/.ipython',
 '/home/mz/.local/lib/python3.8/site-packages',
 '/home/mz/.pyenv/versions/3.8.12/envs/lewagon/lib/python3.8/site-packages/gitdb/ext/smmap']

__sys.version__

This attribute displays a string containing the version number of the current Python interpreter. 

In [361]:
sys.version

'3.8.12 (default, Dec 16 2021, 20:07:31) \n[GCC 9.3.0]'

# error handling

## try ... except

In [39]:
try:
    print(a)
except:
    print('we have an error')

t = '\n the code continues even if there was an error, \n because we skipped the usual error message \n replaced it with our own message and got on with work'
print(t)

we have an error

 the code continues even if there was an error, 
 because we skipped the usual error message 
 replaced it with our own message and got on with work


## try...except...else

In [40]:
try:
    a ='Hello'
    print(a)
except:
    print('we have an error')
else: 
    print('All went well')


hello
all went well


## finally
- get's executed in any case: error or no error

In [46]:
try:
    # a = 'Hello'
    print(d)
except:
    print('we have an error')
finally:
    print('\nI wanted to inform you nonetheless')


we have an error

I wanted to inform you nonetheless


## raise exception

In [48]:
num = 10 
if num > 5:
    raise Exception('num is more than 5')

Exception: num is more than 5

## Exception kind

In [49]:
try: 
    print(x)
except NameError:
    print('You used a variable that is not defined')

you useed a variable that is not defined


# different stuff

## if \_\_name\_\_ == "\_\_main\_\_"
- When the interpreter runs a module, the __name__ variable will be set as  __main__ if the module that is being run is the main program.
- But if the code is importing the module from another module, then the __name__  variable will be set to that module’s name.

In [ ]:
# import openpyxl as xl         # xl is an alias for openpyxl
# from openpyxl.chart import BarChart, Reference

# wb = xl.load_workbook("/home/mz/PyScripts/venv001/data/transactions.xlsx")
# sheet = wb["Sheet1"]          # access Sheet1 case-sensitive
# cell = sheet["a1"]            # coordinate of a cell column/row
# # cell = sheet.cell(1,1)      # the same with cell() method
# # print(cell.value)           # value of cell a1
# # print(sheet.max_row)        # how many rows?


## Reg Ex
__findall__ 	Returns a list containing all matches</br>
__search__ 	Returns a Match object if there is a match anywhere in the string</br>
__split__ 	Returns a list where the string has been split at each match</br>
__sub__ 	Replaces one or many matches with a string</br>
https://www.w3schools.com/python/python_regex.asp</br>
https://regex101.com/

In [18]:
import re
txt = "The rain in Spain"
x = re.findall("ai", txt)
print(x)


['ai', 'ai']


## decorators
- Add functionality to an existing function with decorators. This is called metaprogramming.
- A function can take a function as argument (the function to be decorated) and return the same function with or without extension.

In [422]:
# hello() is a decorator function, It wraps the function in the other function.
def hello(func):
    def inner():
        print("Hello ")
        func()
    return inner

#  name() is decorated by the function hello().
def name():
    print("Marco")

obj = hello(name)  # name is decorated/ feed to hello
obj()


Hello 
Marco


In [427]:
# when we want to decorate name with hello we can use this cleaner decoration with @
@hello                                                                                                      
def surname():
    print("Müller")   
    
surname()

Hello 
Müller


In [429]:
import time

# timer function
def measure_time(func):
    def wrapper(*arg):
        t = time.time()
        res = func(*arg)
        print("Function took " + str(time.time() - t) + " seconds to run")
        return res
    return wrapper

# decorate function with the timer fct.
@measure_time
def myFunction(n):
    time.sleep(n)

myFunction(2)


Function took 2.002239227294922 seconds to run


In [ ]:
# It’s critical to emphasize that decorators generally do not alter the calling signature or return value of function being wrapped. 
# The use of *args and**kwargs is there to make sure that any input arguments can be accepted. 
# The return value of a decorator is almost always the result of calling func(*args, **kwargs), where func is the original unwrapped function.

import time


def timeis(func):
    '''Decorator that reports the execution time.'''

    def wrap(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()

        print(func.__name__, end - start)
        return result

    return wrap


@timeis
def countdown(n):
    '''Counts down'''
    while n > 0:
        n -= 1


countdown(5)
countdown(1000)


## datetime
- to convert a string to a date object.

In [14]:
from datetime import datetime
from datetime import date

input_str = '21/01/24 11:04:19'
dt_object = datetime.strptime(input_str, '%d/%m/%y %H:%M:%S')

print("The type of the input date string now is: ", type(dt_object))
print("The date is", dt_object)

print(date.today())
birthday = date(1984, 7, 25)
print(date.today() - birthday)

The type of the input date string now is:  <class 'datetime.datetime'>
The date is 2024-01-21 11:04:19
2022-04-27
13790 days, 0:00:00


In [464]:
print(time.strptime.__doc__)

strptime(string, format) -> struct_time

Parse a string to a time tuple according to a format specification.
See the library reference manual for formatting codes (same as
strftime()).

Commonly used format codes:

%Y  Year with century as a decimal number.
%m  Month as a decimal number [01,12].
%d  Day of the month as a decimal number [01,31].
%H  Hour (24-hour clock) as a decimal number [00,23].
%M  Minute as a decimal number [00,59].
%S  Second as a decimal number [00,61].
%z  Time zone offset from UTC.
%a  Locale's abbreviated weekday name.
%A  Locale's full weekday name.
%b  Locale's abbreviated month name.
%B  Locale's full month name.
%c  Locale's appropriate date and time representation.
%I  Hour (12-hour clock) as a decimal number [01,12].
%p  Locale's equivalent of either AM or PM.

Other codes may be available on your platform.  See documentation for
the C library strftime function.



## compile()
- If the Python code is in string form or is an AST object, and you want to change it to a code object, then you can use compile() method.
- The code object returned by the compile() method can later be called using methods like: exec() and eval() which will execute dynamically generated Python code.

compile(source, filename, mode, flags=0, dont_inherit=False, optimize=-1)

- Source – It can be a normal string, a byte string, or an AST object
- Filename -This is the file from which the code was read. If it wasn’t read from a file, you can give a name yourself.
- Mode – Mode can be exec, eval or single.
  - a. eval – If the source is a single expression.
  - b. exec – It can take a block of a code that has Python statements, class and functions and so on.
  - c. single – It is used if consists of a single interactive statement


In [ ]:
# Creating sample sourcecode to multiply two variables
# x and y.
srcCode = 'x = 10\ny = 20\nmul = x * y\nprint("mul =", mul)'

# Converting above source code to an executable
execCode = compile(srcCode, 'mulstring', 'exec')

# Running the executable code.
exec(execCode)

## caching 

## ord() & chr - Char to Int and back
- chr() is used for converting an Integer to a Character, while the function 
- ord() is used to do the reverse, i.e, convert a Character to an Integer.

In [ ]:
def character_range(a, b):
    for char in range(ord(a), ord(b) + 1):
        yield (char)

print(list(character_range("a", "z")))
unicode = [chr(x) for x in character_range("a", "z")]  # one-character string of an integer code point.
print(unicode)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122]
['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']


In [ ]:
print(chr(65))  # one-character string of an integer code point.
print(bytearray(4))  # The bytearray() method returns a bytearray object which is an array of the given bytes.
print(bytes(4))  # Returns a bytes object

65
A
bytearray(b'\x00\x00\x00\x00')
b'\x00\x00\x00\x00'


## Keywords 

In [ ]:
import keyword
print(keyword.kwlist)
print(help("keywords"))
help("raise")

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']

Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 

None
The "raise" statement
***************

## file handling - open()
- The open() function takes two parameters; filename, and mode.
-  There are four different methods (modes) for opening a file:
- "r" - Read - Default value. Opens a file for reading, error if the file does not exist
-  "a" - Append - Opens a file for appending, creates the file if it does not exist
-  "w" - Write - Opens a file for writing, creates the file if it does not exist

In [ ]:
with open('unix console.txt', 'r', encoding='utf-8') as fp:
    # read sample.txt
    print(fp.read())

pwd     # parent working dir
ls      # list
cd      # change dir

ls -a   # show hidden .files too
ls -la  # long form and hidden

# relative path
.    # current dir
..   # parent dir



## Path and listing files and directories

In [ ]:
from pathlib import Path
path = Path("ecommerce")    # relative path start from current directory
p = Path("/home/mz/PyScripts/venv001/")    # absolute Path: usr/local/bin # starts from root
print(path.exists())
path = Path("emails")
path.mkdir()              # create a directory
print(Path("emails"))
print(path.rmdir())       # when your in the dir erase it
path = Path()               # go to current dir
for file in path.glob ("*.txt"):
   print(file)


## glob 
- module provides a function for making file lists from directory wildcard searches:

In [ ]:
import glob
print(glob.glob('*.ipynb'))


['Math_Statistics.ipynb', 'Numpy.ipynb', '01_Python_Basics.ipynb', 'functions.ipynb', 'Pandas.ipynb', 'standard_library_tour.ipynb', 'np_pd_plt_sns_intro.ipynb', '02_Python_Advanced.ipynb', 'automate_stuff.ipynb']


## os

In [ ]:
print(dir(os))                  # list of all module functions

In [ ]:
print(help(os))               # extensive manual page created from the module's docstrings

Help on module os:

NAME
    os - OS routines for NT or Posix depending on what system we're on.

MODULE REFERENCE
    https://docs.python.org/3.8/library/os
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This exports:
      - all functions from posix or nt, e.g. unlink, stat, etc.
      - os.path is either posixpath or ntpath
      - os.name is either 'posix' or 'nt'
      - os.curdir is a string representing the current directory (always '.')
      - os.pardir is a string representing the parent directory (always '..')
      - os.sep is the (or a most common) pathname separator ('/' or '\\')
      - os.extsep is the extension separator (always '.')
      - os.altsep is the alternate pathname se

In [ ]:
import os

#Returns the current working directory.
print(os.getcwd())
# Return the real group id of the current process.
print(os.getgid())

/home/mz/code/Stuff/pyfiles
1000


In [ ]:
# Return the current process’s user id.
print(os.getuid())
# Returns the real process ID of the current process.
print(os.getpid())
# Set the current numeric umask and return the previous umask.
print(os.umask)
# Return information identifying the current operating system.
print(os.uname())


1000
12550
<built-in function umask>
posix.uname_result(sysname='Linux', nodename='LAPTOP-FERQ4ACR', release='5.10.16.3-microsoft-standard-WSL2', version='#1 SMP Fri Apr 2 22:23:49 UTC 2021', machine='x86_64')


In [ ]:
# Change the root directory of the current process to path.
print(os.chroot(path))
# Return a list of the entries in the directory given by path.
print(os.listdir(path))
# Create a directory named path with numeric mode mode.
print(os.mkdir(path))
# Recursive directory creation function.
os.makedirs(path)
# Remove (delete) the file path.
os.remove(path)
# Remove directories recursively.
os.removedirs(path)

In [ ]:
# Rename the file or directory src to dst.
os.rename(src, dst)
# Remove (delete) the directory path.
os.rmdir(path)

NameError: name 'src' is not defined

## sys()

In [ ]:
# search path is accessible in the Python variable sys.path
import sys

print(sys.path)
# you can put the module file in any directory of your choice and then modify
# sys.path at run-time so that it contains that directory.
sys.path.append(r'C:\Users\john')


In [ ]:
# The sys module also has attributes for stdin, stdout, and stderr. The latter is useful for emitting
# warnings and error messages to make them visible even when stdout has been redirected:
import sys
sys.stderr.write('Warning, log file not found starting a new one\n')


Warning, log file not found starting a new one


SystemExit: 

/home/mz/.pyenv/versions/3.8.12/envs/lewagon/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3465: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [ ]:
# The most direct way to terminate a script is to use
sys.exit()

SystemExit: 

## timeit - performance measurement
- timeit module quickly demonstrates a modest performance advantage
- https://www.guru99.com/timeit-python-examples.html

In [ ]:
import timeit

import_module = "import random"
testcode = ''' 
def test(): 
    return random.randint(10, 100)
'''
print(timeit.timeit(stmt=testcode, setup=import_module))


0.08800524700200185


# testing
- __UNIT TESTING__ is a type of software testing where individual units or components of a software are tested. 
- The __doctest__ module searches for pieces of text that look like interactive Python sessions, and then executes those sessions to verify that they work exactly as shown. There are several common ways to use doctest:

    - To check that a module’s docstrings are up-to-date by verifying that all interactive examples still work as documented.

    - To perform regression testing by verifying that interactive examples from a test file or a test object work as expected.

    - To write tutorial documentation for a package, liberally illustrated with input-output examples. Depending on whether the examples or the expository text are emphasized, this has the flavor of “literate testing” or “executable documentation”.


In [ ]:
import unittest


In [ ]:
# quality control -> write tests for each function as it is developed
import doctest


## data compression
- Common data archiving and compression formats are directly supported by modules
- including: zlib, gzip, bz2, lzma, zipfile and tarfile.


In [ ]:
import zlib
s = b'witch which has which witches wrist watch'
print(len(s))
t = zlib.compress(s)
print(t)
print(len(t))
print(zlib.decompress(t))
print(zlib.crc32(s))


## doctest & unittest
https://mattermost.com/blog/testing-python-understanding-doctest-and-unittest/